samtranslator/utils/traverse.py

```diff
@@ -0,0 +1,104 @@
+from typing import Any, Dict, List
+
+from samtranslator.utils.resolve_actions import ResolveAction
+
+
+def traverse(template: Dict[str, Any], actions: List[ResolveAction]) -> Any:
+    """
+    Calls the correct traversal method
+
+    :param template: The template that needs modifying
+    :param actions: That actions that need to be performed to the template
+    :return: Modified Template
+
+    TODO: Add multiple strategies and choose between those with an Optional value
+    """
+    return _preorder_traverse(template, actions)
+
+
+def _execute_actions(template: Dict[str, Any], actions: List[ResolveAction]) -> Any:
+    """
+    Calls the resolver methods when they are in the list of ResolveActions. Then calls the actions passed to traverse.
+
+    :param template: The template that needs modifying
+    :param actions: actions to be executed.
+    :return: Modified Template
+    """
+    if len(actions) == 0:
+        return template
+
+    for action in actions:
+        action.execute(template)
+
+    return template
+
+
+def _preorder_traverse(
+    input_value: Any,
+    actions: List[ResolveAction],
+) -> Any:
+    """
+    Driver method that performs the actual traversal of input and calls the appropriate `resolver_method` when
+    to perform the resolution.
+
+    :param input_value: Any primitive type  (dict, array, string etc) whose value might contain a changed value
+    :param actions: Method that will be called to actually resolve the function.
+    :return: Modified `input` with values resolved
+    """
+
+    #
+    # Traversal Algorithm:
```

Comment:
This should go into doc string. And also we should speak about traversing in general instead of resolving